reference/test/matrix/batch_ell_kernels.cpp

```diff
@@ -0,0 +1,433 @@
+    auto alpha = gko::batch_initialize<Vec>({{-1.0}, {1.0}}, this->exec);
+    auto beta = gko::batch_initialize<Vec>({{2.0}, {2.0}}, this->exec);
+    auto x = gko::batch_initialize<Vec>({{2.0, 1.0, 4.0}, {-2.0, 1.0, 4.0}},
+                                        this->exec);
+    auto y = gko::batch_initialize<Vec>({{1.0, 2.0}, {1.0, -2.0}}, this->exec);
+    auto umats = this->mtx->unbatch();
+    auto umtx0 = umats[0].get();
+    auto umtx1 = umats[1].get();
+    auto ualphas = alpha->unbatch();
+    auto ualpha0 = ualphas[0].get();
+    auto ualpha1 = ualphas[1].get();
+    auto ubetas = beta->unbatch();
+    auto ubeta0 = ubetas[0].get();
+    auto ubeta1 = ubetas[1].get();
+    auto uxs = x->unbatch();
+    auto ux0 = uxs[0].get();
+    auto ux1 = uxs[1].get();
+    auto uys = y->unbatch();
+    auto uy0 = uys[0].get();
+    auto uy1 = uys[1].get();
+
+    this->mtx->apply(alpha.get(), x.get(), beta.get(), y.get());
+    umtx0->apply(ualpha0, ux0, ubeta0, uy0);
+    umtx1->apply(ualpha1, ux1, ubeta1, uy1);
+
+    EXPECT_EQ(y->at(0, 0), uy0->at(0));
+    EXPECT_EQ(y->at(0, 1), uy0->at(1));
+    EXPECT_EQ(y->at(1, 0), uy1->at(0));
+    EXPECT_EQ(y->at(1, 1), uy1->at(1));
+}
+
+
+TYPED_TEST(BatchEll, AppliesLinearCombinationToDenseMatrix)
+{
+    using Vec = typename TestFixture::Vec;
+    using T = typename TestFixture::value_type;
+    auto alpha = gko::batch_initialize<Vec>({{1.0}, {-1.0}}, this->exec);
+    auto beta = gko::batch_initialize<Vec>({{2.0}, {-2.0}}, this->exec);
+    auto x = gko::batch_initialize<Vec>(
+        {{I<T>{2.0, 3.0}, I<T>{1.0, -1.5}, I<T>{4.0, 2.5}},
+         {I<T>{2.0, 2.0}, I<T>{-1.0, -1.5}, I<T>{4.0, 2.5}}},
+        this->exec);
+    auto y = gko::batch_initialize<Vec>(
+        {{I<T>{1.0, 0.5}, I<T>{2.0, -1.5}}, {I<T>{2.0, 1.5}, I<T>{2.0, 1.5}}},
+        this->exec);
+
+    auto umats = this->mtx->unbatch();
+    auto umtx0 = umats[0].get();
+    auto umtx1 = umats[1].get();
+    auto ualphas = alpha->unbatch();
+    auto ualpha0 = ualphas[0].get();
+    auto ualpha1 = ualphas[1].get();
+    auto ubetas = beta->unbatch();
+    auto ubeta0 = ubetas[0].get();
+    auto ubeta1 = ubetas[1].get();
+    auto uxs = x->unbatch();
+    auto ux0 = uxs[0].get();
+    auto ux1 = uxs[1].get();
+    auto uys = y->unbatch();
+    auto uy0 = uys[0].get();
+    auto uy1 = uys[1].get();
+
+    this->mtx->apply(alpha.get(), x.get(), beta.get(), y.get());
+    umtx0->apply(ualpha0, ux0, ubeta0, uy0);
+    umtx1->apply(ualpha1, ux1, ubeta1, uy1);
+
+    EXPECT_EQ(y->at(0, 0, 0), uy0->at(0, 0));
+    EXPECT_EQ(y->at(0, 1, 0), uy0->at(1, 0));
+    EXPECT_EQ(y->at(0, 0, 1), uy0->at(0, 1));
+    EXPECT_EQ(y->at(0, 1, 1), uy0->at(1, 1));
+    EXPECT_EQ(y->at(1, 0, 0), uy1->at(0, 0));
+    EXPECT_EQ(y->at(1, 1, 0), uy1->at(1, 0));
+    EXPECT_EQ(y->at(1, 0, 1), uy1->at(0, 1));
+    EXPECT_EQ(y->at(1, 1, 1), uy1->at(1, 1));
+}
+
+
+TYPED_TEST(BatchEll, ApplyFailsOnWrongInnerDimension)
```

Comment:
I guess these kinds of tests should also be in `core/test/matrix/batch_ell.cpp`?

Reply:
I guess technically yes, but we do keep it in reference in other non-batched kernels, as apply is not technically a core function, in the sense that it needs to call a kernel.

Reply:
Now that reference is always built, and the actual thing being tested here is in core, I think it should go to core. But if you want to leave it here, that's fine.